NIFI-3255 removed dependency on session.merge from SplitText

olegz opened the pull request:
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

ijokarumawak commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java:
```diff
@@ -326,6 +329,27 @@ public void process(InputStream in) throws IOException {
+     * sessioin.merge. Also, it always expects 2 and no more then 2 flow files,
+     * so vararg signature primarily for simplification.
+     */
+    private FlowFile merge(ProcessSession session, FlowFile... flowFiles) {
```
Since this method does simply adding header to each split, naming this as `prependHeader` would make its purpose clearer. The keyword `merge` is counter-intuitive in the context of this SplitText processor. How do you think a method signature like this?

```java
/**
 * @param split a FlowFile for the split, this will be removed from session
 * @return newly created flow file contains header and split merged
 **/
private FlowFile prependHeader(ProcessSession session, FlowFile header, FlowFile split)
```

With this, we don't have to mention that this is different from session.merge, nor explain about the var args.

olegz replied:
So, in the context of this specific processor what you're saying is true - _prepend header_. But, the operation itself has nothing to do with either _headers_ or _prepend_.  It simply merges two files together in the order they have been supplied and eventually should probably be considered as candidate for some utility static operation. Anyway, I'll keep the name as is, but will update the javadoc.

ijokarumawak commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java:
```diff
@@ -326,6 +329,27 @@ public void process(InputStream in) throws IOException {
 
+    /**
+     * Will merge the two files. This operation has different semantics then
+     * sessioin.merge. Also, it always expects 2 and no more then 2 flow files,
```
Typo `sessioin`. If we have cycle to update this PR, I'd prefer having this method suggested as below.

ijokarumawak commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java:
```diff
@@ -326,6 +329,27 @@ public void process(InputStream in) throws IOException {
+     * so vararg signature primarily for simplification.
+     */
+    private FlowFile merge(ProcessSession session, FlowFile... flowFiles) {
+        FlowFile mergedFlowFile = session.create();
```
Please pass the source FlowFile to session.create so that it generates fork provenance event. Calling create() without that loses track to the parent provenance.

olegz replied:
Koji, as you can see, there are no source/target files here. Just an array of FlowFiles that need to be concatenated in the order they were provided. It is a simple internal utility operation which requires no provenance. 
The fact that I am storing header information in what essentially is a temporary FlowFile, does not mean I am deriving A from B.

ijokarumawak replied:
@olegz Excuse me if I wasn't clear enough, but I was mentioning about the relationship between original incoming FlowFile and outgoing FlowFiles those are derived from the original FlowFile.

The 2nd one in the var arg is a FlowFile which is derived from the original and it has provenance to the original file, but if we create another FlowFile here, the lineage gets lost.

ijokarumawak commented:
Hi @olegz have you had a chance to see my review comments? The issue not being able to generate fork provenance should be addressed. Thanks!

olegz commented:
@ijokarumawak sorry, didn't get a chance to get to it yet. Will definitely address.

olegz commented:
@ijokarumawak addressed some of the PR comments and commented on the others. Changed the name of the _merge(..)_ operation to _concatenateContents(..)_.
Pushed. . .

ijokarumawak commented:
@olegz Thanks for updating the PR. Please check the comment I put on the provenance once again.

olegz commented:
@ijokarumawak addressing it now and will merge. So basically I will be creating splits via ```FlowFile mergedFlowFile = session.create(sourceFlowFile);``` passing the source file from ```onTrigger``` to create appropriate linkage.
